[do not merge] Versions, Deployments, Gradual Deployments (#13429)
cloudflare/cloudflare-docs: content/workers/configuration/versions-and-deployments/_index.md

---
pcx_content_type: concept

> title: Versions & Deployments

Reviewer: Why one page rather than distinct pages?

Author: The concepts are so coupled, thought it would make more sense as one page. Esp as we get into how to use the two together. Open to splitting up but this page flows well IMO.

meta:
  description: Upload versions of Workers and create deployments to release new versions. 
---

> {{<heading-pill style="beta">}}Versions & Deployments{{</heading-pill>}}

Reviewer: Something visual to provide?

Author: Good idea. Visuals coming!

## Versions
Versions track the state of code as well as the state of configuration defined in a script’s wrangler.toml file. They track historical changes to [bundled code](/workers/wrangler/bundling/) and changes to configuration like [bindings](/workers/configuration/bindings/), [compatibility date and compatibility flags](/workers/configuration/compatibility-dates/) over time.

{{<Aside type="note">}}

State changes for associated Workers storage resources such as [KV](/kv/), [R2](/r2/), [Durable Objects](/durable-objects/) and [D1](/d1/) are not tracked with versions.

{{</Aside>}}

## Deployments

Deployments track the version(s) of your Worker that are actively serving traffic. 

By default, changes to Workers through `npx wrangler deploy`, via the Cloudflare dashboard or the [Workers Script Upload API](https://developers.cloudflare.com/api/operations/worker-script-upload-worker-module) automatically create a new version that is deployed to 100% of traffic.

## Using Versions and Deployments

> Together, versions and deployments are used to track changes to your Worker and configure how those changes are deployed to your traffic. With versions and deployments, it is possible to upload changes to your Worker independent of changing the version that's actively serving traffic. This is useful if:

Reviewer: Everything in docs should be like you're talking directly to a person and active voice

```suggestion
Together, you can use versions and deployments to track changes to your Worker and configure how those changes are deployed to your traffic. With versions and deployments, you can upload changes to your Worker independent of changing the version that's actively serving traffic. This is useful if:
```

{{</Aside>}}

{{<Aside type="note">}}

**New versions are not triggered by changes to bound resources.**

For example, if two Workers (Worker A and Worker B) are bound via a [service binding](/workers/configuration/bindings/about-service-bindings/), changing the code of Worker B will not trigger a new version of Worker A. It will only create a new version of Worker B. Changes to the service binding (such as, deleting the binding or updating the [environment](/workers/wrangler/environments/) it points to) on Worker A will also not create a new version for Worker B.

{{</Aside>}}

## Viewing versions and deployments

### Via Wrangler

Wrangler allows you to view the 20 most recent versions and deployments.

Refer to the [`versions` and `deployments` command documentation](/workers/wrangler/commands#deployments) to view commands. 

### Via the Cloudflare dashboard

**To view your deployments:**

1. Log in to the [Cloudflare dashboard](https://dash.cloudflare.com) and select your account.
2. Select **Workers & Pages**.

> 3. In **Overview**, select your Worker > **Deployments**. 

Reviewer: Would just link directly to Workers section of dash using `https://dash.cloudflare.com/?to=/:account/workers` format

Your active deployment includes the version(s) that are currently serving traffic. To view previous deployments, select 'View all deployments'. 

Deployments metadata includes information about the versions in the deployment, when the deployment was created, the author, source and deployment message.  

**To view your versions:**

1. Log in to the [Cloudflare dashboard](https://dash.cloudflare.com) and select your account.
2. Select **Workers & Pages**.
3. In **Overview**, select your Worker > **Deployments**. 

Version history tracks the state of code and config for a Worker. Versions metadata includes information about when the deployment was created, the author, source, version message and version tag.  

{{<Aside type="note">}}

> Deployments are in active development. To give feedback, request a [live chat](https://www.cloudflare.com/lp/developer-week-deployments).

Reviewer: A live chat?

---
pcx_content_type: concept
title: Versions & Deployments
meta:
  description: Upload versions of Workers and create deployments to release new versions. 
---

{{<heading-pill style="beta">}}Versions & Deployments{{</heading-pill>}}

Versions and deployments are used to track changes to your Worker and configure how those changes are deployed to your traffic.

{{<Aside type="note">}}

Versions and deployments are in **beta and under active development**. Please read [limitations](/workers/configuration/versions-and-deployments/#limitations) before using these features.

Provide your feeback through this [feedback form](https://www.cloudflare.com/lp/developer-week-deployments).

{{</Aside>}}

## Versions

> Versions track the state of code as well as the state of configuration defined in a script’s [wrangler.toml](/workers/wrangler/configuration/) file. They track historical changes to [bundled code](/workers/wrangler/bundling/) and changes to configuration like [bindings](/workers/configuration/bindings/), [compatibility date and compatibility flags](/workers/configuration/compatibility-dates/) over time.

Reviewer 2: [nit] Only minor but I wonder if using the term "track" is a little vague here.

I'm not sure I have much better suggestions, but you could alternatively something like "each version is defined by/composed of code and configuration etc" instead or use that phrasing to elaborate on what you mean by versions "tracking" the state of code and configuration etc.

Metadata associated with a version is also tracked, including the version ID, user that created the version, deploy source, timestamp. Optionally, a version message and version tag can be configured on version upload. 

{{<Aside type="note">}}

State changes for associated Workers storage resources such as [KV](/kv/), [R2](/r2/), [Durable Objects](/durable-objects/) and [D1](/d1/) are not tracked with versions.

{{</Aside>}}

## Deployments

Deployments track the version(s) of your Worker that are actively serving traffic. A deployment can consist of one or two versions of a Worker. 

> By default, Workers supports an all-at-once deployment model where traffic is immediatley shifted from one version to the newly deployed version automatically. Alternatively, [gradual deployments]((/gradual-deployments)) can be used to safely roll out a new version by incrementally shifting traffic it.

Reviewer 3: small grammatical thing, should be "... shifting traffic to it."

Metadata associated with a deployment is also tracked, including the user that created the deployment, deploy source, timestamp and the version(s) in the deployment. Optionally, a deployment message can be configured when a deployment is created. 

## Using Versions and Deployments

With versions and deployments, you can upload changes to your Worker independent of changing the version that's actively serving traffic. This is useful if:

- You're running critical applications on Workers and want to reduce risk when deploying new versions of your Worker.
- You want to monitor for performance differences when deploying new versions of your Worker.
- You have a CI pipeline configured for Workers, but want to cut manual releases.

### New versions are created one of two ways

**Uploading a new version and deploying it immediately**

> Changes uploaded with `npx wrangler deploy`, via the Cloudflare dashboard or the [Workers Script Upload API](https://developers.cloudflare.com/api/operations/worker-script-upload-worker-module) create a new version that is automatically deployed to 100% of traffic. 

Reviewer 4: ```suggestion
Changes uploaded with [`wrangler deploy`](/workers/wrangler/commands/#deploy), via the Cloudflare dashboard or the [Workers Script Upload API](https://developers.cloudflare.com/api/operations/worker-script-upload-worker-module) create a new version that is automatically deployed to 100% of traffic. 
```

**Uploading a new version to be [gradually deployed](/gradual-deployments) or deployed at a later time** 

> In order to create a new version that is not deployed immediately, use the new wrangler command `npx wrangler versions upload --experimental-gradual-rollouts` or create a new version via the Cloudflare dashboarding using the "Save changes" buttons.

Reviewer 4: ```suggestion
In order to create a new version that is not deployed immediately, you must use [gradual deployments](/gradual-deployments). To use gradual deployments, use the [`wrangler versions upload --experimental-gradual-rollouts`](/workers/wrangler/commands/#upload) command or create a new version via the Cloudflare dashboarding by selecting **Save changes** on your Worker.
```

> For more details on the wrangler commands, refer to the [`versions` documentation](/workers/wrangler/commands/#versions). 

Reviewer 4: ```suggestion
For more details on the wrangler commands, refer to the [`wrangler versions` command documentation](/workers/wrangler/commands/#versions). 
```

{{<Aside type="note">}}

> **New versions are not triggered by changes from [resources connected to your Worker](/workers/runtime-apis/bindings/)**

Reviewer 4: ```suggestion
**New versions are not triggered by changes from [resources connected to your Worker](/workers/configuration/bindings/).**
```

For example, if two Workers (Worker A and Worker B) are connected via a [service binding](/workers/configuration/bindings/about-service-bindings/), changing the code of Worker B will not create a new version of Worker A. It will only create a new version of Worker B. Changes to the service binding (such as, deleting the binding or updating the [environment](/workers/wrangler/environments/) it points to) on Worker A will also not create a new version for Worker B.

{{</Aside>}}

## Viewing versions and deployments

### Via Wrangler

> Wrangler allows you to view the 10 most recent versions and deployments.Refer to the [`versions`](/workers/wrangler/commands#versions) and [`deployments`](/workers/wrangler/commands/#deployments) documentation to view the commands. 

Reviewer 4: ```suggestion
Wrangler allows you to view the 10 most recent versions and deployments. Refer to the [`versions`](/workers/wrangler/commands#versions) and [`deployments`](/workers/wrangler/commands/#deployments) documentation to view the commands. 
```

### Via the Cloudflare dashboard

> 1. Log in to the [Cloudflare dashboard](https://dash.cloudflare.com/?to=/:account/workers) and select your account.

Reviewer 4: ```suggestion
To view your deployments in the Cloudflare dashboard:

1. Log in to the [Cloudflare dashboard](https://dash.cloudflare.com/?to=/:account/workers) and select your account.
```

2. Select your Worker > **Deployments**.  

> ## Limitations

Reviewer: include service workers bit

> **Making changes to Routes, Custom Domains, Cron Triggers, Logpush or Tail Workers**

Reviewer 4: ```suggestion
### Routes, Custom Domains, Cron Triggers, Logpush, and Tail Workers
```

Author: Want to be specific here that that limitation applies when making changes/updates to these settings. Leaving as is.

> As of now, wrangler users must deploy upates to [Routes](/workers/configuration/routing/routes/), [Custom Domains](/workers/configuration/routing/custom-domains/), [Cron Triggers](/workers/configuration/cron-triggers/), [Logpush](/observability/logging/logpush/) or [Tail Workers](observability/logging/tail-workers/) using `npx wrangler deploy`. Updates to this configuration using `npx wrangler versions upload --experimental-gradual-rollouts` will not take effect. This will be fixed in the near future. 

Reviewer 4: ```suggestion
As of now, Wrangler users must deploy upates to [Routes](/workers/configuration/routing/routes/), [Custom Domains](/workers/configuration/routing/custom-domains/), [Cron Triggers](/workers/configuration/cron-triggers/), [Logpush](/observability/logging/logpush/) or [Tail Workers](observability/logging/tail-workers/) using `wrangler deploy`. Updates to routes, Custom Domains, Cron Triggers, Logpush, and Tail Workers using `npx wrangler versions upload --experimental-gradual-rollouts` will not take effect. This will be fixed in the near future. 
```

> As of now, wrangler users must deploy updates to [Routes](/workers/configuration/routing/routes/), [Custom Domains](/workers/configuration/routing/custom-domains/), [Cron Triggers](/workers/configuration/cron-triggers/), [Logpush](/observability/logging/logpush/) or [Tail Workers](observability/logging/tail-workers/) using `npx wrangler deploy`. Updates to this configuration using `npx wrangler versions upload --experimental-gradual-rollouts` will not take effect. This will be fixed in the near future. 

Reviewer 4: ```suggestion
Currently, Wrangler users must deploy updates to [routes](/workers/configuration/routing/routes/), [Custom Domains](/workers/configuration/routing/custom-domains/), [Cron Triggers](/workers/configuration/cron-triggers/), [Logpush](/observability/logging/logpush/) or [Tail Workers](observability/logging/tail-workers/) using `npx wrangler deploy`. Updates to routes, Custom Domains, Cron Triggers, Logpush, and Tail Workers using `npx wrangler versions upload --experimental-gradual-rollouts` will not take effect. This will be fixed in the near future. 
```

By default, Workers supports an all-at-once deployment model where traffic is immediately shifted from one version to the newly deployed version automatically. Alternatively, you can use [gradual deployments](/gradual-deployments) to create a rolling deployment strategy by incrementally shifting traffic a new version of the Worker.

You can also track metadata associated with a deployment, including: the user that created the deployment, deploy source, timestamp and the version(s) in the deployment. Optionally, you can configure a deployment message when you create a deployment. 

## Use versions and deployments

### Create a new version

Review the different ways you can create versions of your Worker and deploy them.

#### Upload a new version and deploy it immediately

Changes uploaded with [`wrangler deploy`](/workers/wrangler/commands/#deploy), via the Cloudflare dashboard, or the [Workers Script Upload API](/api/operations/worker-script-upload-worker-module) create a new version that is automatically deployed to 100% of traffic. 

#### Upload a new version to be gradually deployed or deployed at a later time 

> To create a new version of your Worker that is not deployed immediately, use the [`wrangler versions upload --experimental-versions`](/workers/wrangler/commands/#upload) command or create a new version via the Cloudflare dashboard using the **Save changes** button.

Reviewer 4: Save changes button... in the Quick Editor? Need to flesh this out a bit more. Save changes button is not a top-level action in UI.

---
pcx_content_type: concept
title: Versions & Deployments
meta:
  description: Upload versions of Workers and create deployments to release new versions. 
---

{{<heading-pill style="beta">}}Versions & Deployments{{</heading-pill>}}

Versions track changes to your Worker. Deployments configure how those changes are deployed to your traffic. 

You can upload changes (versions) to your Worker independent of changing the version that is actively serving traffic (deployment). Using versions and deployments is useful if:

- You are running critical applications on Workers and want to reduce risk when deploying new versions of your Worker.
- You want to monitor for performance differences when deploying new versions of your Worker.
- You have a CI pipeline configured for Workers but want to cut manual releases.

{{<Aside type="note">}}

Versions and deployments are in **beta and under active development**. Refer to [Limits](/workers/configuration/versions-and-deployments/#limits) before using these features.

Provide your feeback through the [feedback form](https://www.cloudflare.com/lp/developer-week-deployments).

{{</Aside>}}

## Versions

> A version is defined by the state of code as well as the state of configuration defined in a Worker's [`wrangler.toml`](/workers/wrangler/configuration/) file. Versions track historical changes to [bundled code](/workers/wrangler/bundling/) and changes to configuration like [bindings](/workers/configuration/bindings/) and [compatibility date and compatibility flags](/workers/configuration/compatibility-dates/) over time.

Reviewer: ```suggestion
A version represents a Worker's code and [configuration](/workers/wrangler/configuration/) at a point in its history. Versions track historical changes to [bundled code](/workers/wrangler/bundling/) and changes to configuration like [bindings](/workers/configuration/bindings/) and [compatibility date and compatibility flags](/workers/configuration/compatibility-dates/) over time.
```

Worth wordsmithing this 1-2 sentence and getting thing that can be repeated over and over when we talk about this, to use consistent language.

Versions also track metadata associated with a version, including: the version ID, the user that created the version, deploy source, timestamp. Optionally, a version message and version tag can be configured on version upload. 

{{<Aside type="note">}}

State changes for associated Workers storage resources such as [KV](/kv/), [R2](/r2/), [Durable Objects](/durable-objects/) and [D1](/d1/) are not tracked with versions.

{{</Aside>}}

## Deployments

> Deployments track the version(s) of your Worker that are actively serving traffic. A deployment can consist of one or two versions of a Worker. 

Reviewer: Alt — or "currently running on Cloudflare's network". Could see either way, but similarly worth wordsmithing first sentence.

> By default, Workers supports an all-at-once deployment model where traffic is immediately shifted from one version to the newly deployed version automatically. Alternatively, [gradual deployments]((/gradual-deployments)) can be used to safely roll out a new version by incrementally shifting traffic to it.

Reviewer: ```suggestion
By default, Workers are deployed instantly — changes are deployed globally in seconds, and traffic is immediately shifted from the existing version to the newly deployed version automatically. Alternatively, [gradual deployments]((/gradual-deployments)) can be used to safely roll out a new version by incrementally shifting traffic to it.
```

> Metadata associated with a deployment is also tracked, including: the user that created the deployment, deploy source, timestamp and the version(s) in the deployment. Optionally, a deployment message can be configured when a deployment is created. 

Reviewer: "You can also track..."

+ links to how to do this?

## Using Versions and Deployments

### Create a new version

Review the different ways you can create versions of your Worker and deploy them.

#### Upload a new version and deploy it immediately

Changes uploaded with [`wrangler deploy`](/workers/wrangler/commands/#deploy), via the Cloudflare dashboard or the [Workers Script Upload API](https://developers.cloudflare.com/api/operations/worker-script-upload-worker-module) create a new version that is automatically deployed to 100% of traffic. 

> **Uploading a new version to be [gradually deployed](/gradual-deployments) or deployed at a later time** 

Reviewer: Is this supposed to be a heading?

In order to create a new version that is not deployed immediately, use the new wrangler command `npx wrangler versions upload --experimental-versions` or create a new version via the Cloudflare dashboarding using the "Save changes" buttons.

For more details on the wrangler commands, refer to the [`wrangler versions` command documentation](/workers/wrangler/commands/#versions). 

{{<Aside type="note">}}

**New versions are not created when you make changes to [resources connected to your Worker](/workers/runtime-apis/bindings/)**

For example, if two Workers (Worker A and Worker B) are connected via a [service binding](/workers/configuration/bindings/about-service-bindings/), changing the code of Worker B will not create a new version of Worker A. It will only create a new version of Worker B. Changes to the service binding (such as, deleting the binding or updating the [environment](/workers/wrangler/environments/) it points to) on Worker A will also not create a new version of Worker B.

{{</Aside>}}

## View versions and deployments

### Via Wrangler

> Wrangler allows you to view the 10 most recent versions and deployments. Refer to the [`versions`](/workers/wrangler/commands#versions) and [`deployments`](/workers/wrangler/commands/#deployments) documentation to view the commands. 

Reviewer: Think worth making partial so that content can exist directly here as well?
